controller/api/destination/watcher/endpoints_watcher.go

```diff
@@ -404,6 +423,127 @@ func (ew *EndpointsWatcher) getServicePublisher(id ServiceID) (sp *servicePublis
 	return
 }
 
+func (ew *EndpointsWatcher) addServer(obj interface{}) {
+	server := obj.(*v1beta1.Server)
+
+	// Get the set of pods that the server selects so that we can check if a
+	// port publisher exists for that endpoint, and if so it's new set of
+	// opaque pod ports is received.
+	options, err := createListOptions(server.Spec.PodSelector)
+	if err != nil {
+		ew.log.Errorf("failed to create Server ListOptions: %s", err)
+		return
+	}
+	pods, err := ew.k8sAPI.Client.CoreV1().Pods("").List(context.Background(), options)
+	if err != nil {
+		ew.log.Errorf("failed to list pods: %v", err)
+		return
+	}
+
+	// publishersToUpdate tracks the port publishers that should send new
+	// address sets once the new set of opaque pod ports is created for each
+	// one.
+	publishersToUpdate := make(map[*portPublisher]struct{})
+
+	for _, pod := range pods.Items {
+		port := getPortIntFromPod(pod, server.Spec.Port)
+		for _, sp := range ew.publishers {
+			for _, pp := range sp.ports {
+				for _, addr := range pp.addresses.Addresses {
+					if addr.Pod != nil && addr.Pod.Name == pod.Name && addr.Pod.Namespace == pod.Namespace {
+						id := PodID{
+							Name:      addr.Pod.Name,
+							Namespace: addr.Pod.Namespace,
+						}
+						if server.Spec.ProxyProtocol == "opaque" {
+							// The server's protocol is opaque, so we make
+							// sure the pod's matching port is set as opaque
+							// on the address set.
+							if ports, ok := pp.addresses.OpaquePodPorts[id]; ok {
```

Comment:
If we move the opaqueProtocol field to the address struct, I think this logic gets a lot simpler:

```
for _, addr := range pp.addresses.Addresses {
  if addr.Pod != nil && ... {
    if matchesSelector(server, pod) {
      addr.opaqueProtocol = true
    }
  }
}
```